x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,728 @@
+
+## Plan for carrying out migrations of rule SOs
+
+### Migration strategy: perform migration on all write operations of the rules
+
+In order to migrate our users rule's data to the new schema, we should perform the migration of rules whenever any endpoint operation that writes/updates a rule saved object.
+
+> In Elasticsearch, the type of rule saved objects is called `alert`.
+
+This migration strategy of write/update operations means that a user's data will be migrated incrementally, and that both types of data (non-migrated and migrated saved objects) will coexist for an indeterminate amount of time. Therefore we have to maintain backwards data compatibility for the non-migrated data types.
+
+The migration logic should take place as a first step within the handler logic of all endpoints that carry out write/update endpoint operations, and the endpoint should return the already-migrated rule(s).
+
+### Rule Management endpoints that should include this migration logic are:
+
+All of the following endpoints either fetch the rule before updating it, or send the rule's params as part of the body in the request. Therefore, in order to carry out the migration, we can check if the `params` fields already contains the `prebuilt` and `customized` fields, create them within the `params` if they don't, and then proceed to saving them to Elasticsearch.
+
+- **Update Rule** - `PUT /rules`: same as Patch Rule.
+    - Used **only** when updating/modifying a single rule via the Rule Details page
+- **Patch Rule** - `PATCH /rules`: should migrate patched rule to prevent the endpoint from returning an outdated (non-migrated) rule.
+    - Used for attaching shared exceptions list to rules
+- **Bulk Update Rules** - `PUT /rules/_bulk_update`: same as for Patch Rule
+    - Deprecated and unused by the UI (might still be used by API users)
+- **Bulk Patch Rules** - `PATCH /rules/_bulk_update`: same as for Patch Rule
+    - Deprecated and unused by the UI (might still be used by API users)
+- **Import Rules** - `POST /rules/_import`: 
+    - if the user imports rules with the updated rule schema (`prebuilt` and `customized` fields are present), then save rule as-is. 
+    - if the user imports a "legacy" rule (with the `immutable` field, and no `prebuilt` or `customized` fields), the `immutable` field will be saved for backwards compatibility as well, and the `prebuilt` and `customized` fields should always be set to `false` (importing a non-migrated rule means that the rule could not be modified or exported in the Kibana version where it was exported from, so it could not have been a prebuilt rule).
```

Comment:
> if the user imports a "legacy" rule (with the `immutable` field, and no `prebuilt` or `customized` fields)

Does this refer to the `immutable: false` case? Basically, custom rules that our users can export and import right now. Are we talking about backward compatibility with ndjson files containing such rules?

Also, not sure I understand this sentence, could you rephrase?

> importing a non-migrated rule means that the rule could not be modified or exported in the Kibana version where it was exported from, so it could not have been a prebuilt rule

Reply:
> Does this refer to the immutable: false case? Basically, custom rules that our users can export and import right now.

Not only that case, but also rules with `immutable: true` case and non-migrated (`prebuilt` and `customized` do not exist).

> Are we talking about backward compatibility with ndjson files containing such rules?

Yes: we want to handle the case of import of both migrated and non-migrated rules via an ndjson file. I understand that importing rules with `immutable: false` "should" be the only case for non-migrated rules (since `immutable: true` rules cannot be exported), but I think we should also handle the case of users manually modifying the values of rules in the NDJSON file. 
I think that, if a user attempts to import a rule with  `immutable: true` and missing `prebuilt` and `customized` rules, then the most likely scenario is a user trying to import Elastic prebuilt rules from a modfied NDJSON file and we should mark those rules as `prebuilt: true` in ES and attempt to track them to rules in the Fleet package via their `rule_id`.

> Also, not sure I understand this sentence, could you rephrase?

I'm removing that phrase, I thought this over again and the paragraph above makes the most sense to me. I'm writing that idea in the RFC.

Reply:
I moved everything related to them import endpoint to the [Importing rules](#importing-rules){#toc-importing-rules} section and linking there.